### PR TITLE
Document go code snippets as examples

### LINKED ISSUE #204: Document go code snippets as examples
https://godocs.io/testing#hdr-Examples
https://go.dev/blog/examples

As an example of using examples see:
https://pkg.go.dev/io/fs#example-WalkDir

### PR DESCRIPTION
**Pull request by smallhive**
closes #204

### CODE REVIEW

File: accounting/doc.go [hunk 1]
```diff
@@ -1,33 +1,7 @@
...
-
-On server side:
-
-	// recv msg
-
-	var dec accounting.Decimal
-	dec.ReadFromV2(msg)
-
-	// process dec
+Decimal type provides functionality to process user balances.
```
Comment (roman-khimov) [inline accounting/doc.go:4]: `[Decimal]`?

File: accounting/example_test.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
+package accounting_test
+
+import (
+	apiGoAccounting "github.com/nspcc-dev/neofs-api-go/v2/accounting"
+	"github.com/nspcc-dev/neofs-sdk-go/accounting"
+)
+
+// Decimal type provides functionality to process user balances. For example, when
```
Comment (roman-khimov) [inline accounting/example_test.go:8]: Doesn't seem to be a useful comment for this example.

File: accounting/example_test.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+	var dec accounting.Decimal
+	var msg apiGoAccounting.Decimal
+	dec.WriteToV2(&msg)
+
+	// send msg
+}
+
+// Instances can be also used to process NeoFS API V2 protocol messages
+// (see neo.fs.v2.accounting package in https://github.com/nspcc-dev/neofs-api) on server side.
```
Comment (roman-khimov) [inline accounting/example_test.go:31]: Just use a package reference, it should be converted to an appropriate link.

File: accounting/example_test.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+func ExampleDecimal_SetValue() {
+	var val int64
+
+	var dec accounting.Decimal
+	dec.SetValue(val)
+	dec.SetPrecision(8)
+}
+
+// Instances can be also used to process NeoFS API V2 protocol messages
+// (see neo.fs.v2.accounting package in https://github.com/nspcc-dev/neofs-api) on client side.
```
Comment (roman-khimov) [inline accounting/example_test.go:19]: And here as well.

File: bearer/example_test.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+	bearerToken.SetExp(500)
+	bearerToken.SetIat(10)
+	bearerToken.SetNbf(10)
+	bearerToken.SetEACLTable(eaclTable)
+	bearerToken.ForUser(ownerID)
+}
+
+// Bearer token must be signed by owner of the container.
+func ExampleToken_Sign() {
```
Comment (roman-khimov) [inline bearer/example_test.go:33]: This can be merged with `Example()`.

File: pool/example_test.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+	p, _ := pool.New(
+		pool.NewFlatNodeParams([]string{"grpc://localhost:8080", "grpcs://localhost:8081"}),
+		signer,
+		opts,
+	)
+	_ = p
+}
+
+func ExamplePool_ObjectPutInit_explicitAutoSessionDisabling() {
```
Comment (roman-khimov) [inline pool/example_test.go:71]: Doesn't look like an example, too small for it.

File: pool/example_test.go [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+	var prm client.PrmObjectPutInit
+
+	// If you don't provide the session manually with prm.WithinSession, the request will be executed without session.
+	prm.IgnoreSession()
+	// ...
+}
+
+func ExamplePool_ObjectPutInit_autoSessionDisabling() {
```
Comment (roman-khimov) [inline pool/example_test.go:79]: And this one as well.

File: pool/example_test.go [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+	// Auto-session disabled, because you provided session already.
+	prm.WithinSession(sess)
+}
+
+// Connect to the NeoFS server.
+func ExamplePool_Dial() {
+	var p pool.Pool
+
+	// Connect to the NeoFS server
+	_ = p.Dial(context.Background())
```
Comment (roman-khimov) [inline pool/example_test.go:92]: Will fail, there are no nodes configured, so it's not an example.

File: session/example_test.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
+	neofscrypto "github.com/nspcc-dev/neofs-sdk-go/crypto"
+	"github.com/nspcc-dev/neofs-sdk-go/session"
+	"github.com/nspcc-dev/neofs-sdk-go/user"
+)
+
+// Both parties agree on a secret (private session key), the possession of which
+// will be authenticated by a trusted person. The principal confirms his trust by
+// signing the public part of the secret (public session key).
+func ExampleContainer_Sign() {
```
Comment (roman-khimov) [inline session/example_test.go:14]: Maybe just `ExampleContainer`?

File: session/example_test.go [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+	// import neofscrypto "github.com/nspcc-dev/neofs-sdk-go/crypto"
+	// import "github.com/nspcc-dev/neofs-sdk-go/user"
+	// import cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
+
+	// you private key/signer, to prove you are you
+	var principalSigner user.Signer
+	// trusted party, who can do action on behalf of you
```
Comment (roman-khimov) [inline session/example_test.go:21]: Usually from `SessionCreate`. Do we have an example for `SessionCreate`, btw? That'd be more helpful for people.

File: version/example_test.go [hunk 11]
```diff
@@ -0,0 +1,17 @@
+package version_test
+
+import (
+	"github.com/nspcc-dev/neofs-sdk-go/version"
+)
+
+// In most of the cases it will be enough to use the latest supported NeoFS API version in SDK.
+func ExampleCurrent() {
+	_ = version.Current()
```
Comment (roman-khimov) [inline version/example_test.go:9]: Really cool example!

File: version/example_test.go [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+}
+
+// It is possible to specify arbitrary version by setting major and minor numbers.
+func ExampleVersion_SetMajor() {
```
Comment (roman-khimov) [inline version/example_test.go:13]: Doesn't seem to be useful either.

**Review by roman-khimov — COMMENTED** [review]
It'd be great to make our examples more useful. Calling one function doesn't help, examples should show how things interact within the SDK to achieve some end-user goal, what are the prerequisites for some actions and how to properly process results of some calls.

File: version/example_test.go (hunk 11 above)
Comment (smallhive) [reply version/example_test.go]: Removed

File: version/example_test.go (hunk 12 above)
Comment (smallhive) [reply version/example_test.go]: Removed

File: bearer/example_test.go (hunk 5 above)
Comment (smallhive) [reply bearer/example_test.go]: Merged

File: accounting/example_test.go [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+func Example() {
+	var val int64
+
+	var dec accounting.Decimal
+	dec.SetValue(val)
+	dec.SetPrecision(8)
+
+	// Instances can be also used to process NeoFS API V2 protocol messages. See neofs-api-go package.
+
+	// On client side.
```
Comment (roman-khimov) [inline accounting/example_test.go:17]: Either just `Client-side` or `On the client side`.

File: audit/example_test.go [hunk 14]
```diff
@@ -0,0 +1,26 @@
+package audit_test
+
+import (
+	"github.com/nspcc-dev/neofs-sdk-go/audit"
+	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
+)
+
+// Result type groups values which can be gathered during data audit process.
```
Comment (roman-khimov) [inline audit/example_test.go:8]: This should be a `Result` doc.

File: audit/example_test.go [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+func Example() {
```
Comment (roman-khimov) [inline audit/example_test.go:9]: Make it `Example_Result`? The test only covers this structure.

File: client/example_container_put_test.go [hunk 16]
```diff
@@ -1,104 +0,0 @@
...
-	netmapv2 "github.com/nspcc-dev/neofs-api-go/v2/netmap"
-	"github.com/nspcc-dev/neofs-sdk-go/client"
-	"github.com/nspcc-dev/neofs-sdk-go/container"
-	"github.com/nspcc-dev/neofs-sdk-go/container/acl"
-	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
-	"github.com/nspcc-dev/neofs-sdk-go/netmap"
-	"github.com/nspcc-dev/neofs-sdk-go/user"
-)
-
-func ExampleClient_ContainerPut() {
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply client/example_container_put_test.go:18]: I see, but this creates a diff that's not very useful.

File: client/example_test.go [hunk 17]
```diff
@@ -16,9 +16,10 @@ import (
 	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
 	"github.com/nspcc-dev/neofs-sdk-go/netmap"
 	"github.com/nspcc-dev/neofs-sdk-go/user"
+	"github.com/nspcc-dev/neofs-sdk-go/waiter"
 )
 
-func ExampleClient_createInstance() {
+func ExampleNew() {
```
Comment (roman-khimov) [inline client/example_test.go:28]: This rename doesn't belong to 0ca1810bb93075ded7ac8132e4c8763f0dd05875.

File: pool/example_pool_test.go [hunk 18]
```diff
@@ -1,54 +0,0 @@
...
-	"crypto/ecdsa"
-	"crypto/elliptic"
-	"crypto/rand"
-
-	"github.com/nspcc-dev/neofs-sdk-go/client"
-	"github.com/nspcc-dev/neofs-sdk-go/session"
-	"github.com/nspcc-dev/neofs-sdk-go/user"
-)
-
-func ExampleNew_easiestWay() {
```
Comment (roman-khimov) [inline pool/example_pool_test.go:13]: Can we avoid moving them?

File: pool/example_test.go (hunk 6 above)
Comment (roman-khimov) [reply pool/example_test.go:72]: You can add some comments to `ObjectPutInit` call about expected operation results in various cases.

File: pool/example_test.go (hunk 7 above)
Comment (roman-khimov) [reply pool/example_test.go:99]: And here as well (and how it'll differ from other settings).

Comment (smallhive) [reply pool/example_test.go:109]: Added some comments about expected results

File: pool/example_pool_test.go (hunk 18 above)
Comment (roman-khimov) [reply pool/example_pool_test.go:13]: You're mixing real changes with code movements and this makes the result hard to digest. If you're to split these changes into different commits --- I'm OK with that. If you're to just leave old tests the way they are (in separate files) --- even better.

File: pool/example_test.go [hunk 19]
```diff
@@ -0,0 +1,206 @@
...
+	hdr.SetContainerID(cid.ID{})
+	hdr.SetOwnerID(&ownerID)
+
+	var containerID cid.ID
+	// fill containerID
+	var objetID oid.ID
+	// fill objectID
+
+	// In case of a session wasn't provided with prm.WithinSession, the signer must be for account which is a container
+	// owner, otherwise it will be an error.
```
Comment (roman-khimov) [inline pool/example_test.go:100]: `...there will be an error`.

File: pool/example_test.go [hunk 20]
```diff
@@ -0,0 +1,206 @@
...
+
+	// In case of a session was provided with prm.WithinSession, the signer can be ether container owner account or
+	// third party, who can session token, which signer by container owner.
```
Comment (roman-khimov) [inline pool/example_test.go:103]: `...who can use a session token signed by...`?

File: pool/example_test.go [hunk 21]
```diff
@@ -0,0 +1,206 @@
...
+	// acceptable, and it will be an error.
+	// prm.IgnoreSession()
+
+	var ownerID user.ID
+	var hdr object.Object
+	hdr.SetContainerID(cid.ID{})
+	hdr.SetOwnerID(&ownerID)
+
+	// In case of a session wasn't provided with prm.WithinSession, the signer must be for account which is a container
+	// owner, otherwise it will be an error.
```
Comment (roman-khimov) [inline pool/example_test.go:139]: Ditto.

File: pool/example_test.go [hunk 22]
```diff
@@ -0,0 +1,206 @@
...
+
+	// In case of a session was provided with prm.WithinSession, the signer can be ether container owner account or
+	// third party, who can session token, which signer by container owner.
```
Comment (roman-khimov) [inline pool/example_test.go:142]: Ditto.

File: pool/example_pool_test.go (hunk 18 above)
Comment (smallhive) [reply pool/example_pool_test.go:13]: Ok, returned the original file and make two different commits, one - update existing code, the second - move the documentation

File: pool/example_test.go (hunk 19 above)
Comment (smallhive) [reply pool/example_test.go]: updated

File: pool/example_test.go (hunk 20 above)
Comment (smallhive) [reply pool/example_test.go]: updated

File: pool/example_test.go (hunk 21 above)
Comment (smallhive) [reply pool/example_test.go]: updated

File: pool/example_test.go (hunk 22 above)
Comment (smallhive) [reply pool/example_test.go]: updated

File: client/example_container_put_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (smallhive) [reply client/example_container_put_test.go:18]: Oops, missed, reverted client changes as well

**Review by carpawell — COMMENTED** [review]
1. https://github.com/nspcc-dev/neofs-sdk-go/pull/495/commits/8e830554518bfef802d3572662519a0f9f5c2a88: "Example with RCP..."
2. Most of the commits are "remove" but in fact they are "move". But that is cosmetic
3. Really often saw `// send msg`, `// fill container` but it should be `// *send msg*` instead. Not to confuse a reader, IMO

File: audit/example_test.go [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+func ExampleResult() {
+	var res audit.Result
+	var cnr cid.ID
+
+	res.ForEpoch(32)
+	res.ForContainer(cnr)
+	// ...
+	res.Complete()
+
+	// Result instances can be stored in a binary format on reporter side.
```
[the remark this answers is not in the document]
Comment (smallhive) [reply audit/example_test.go]: Updated

File: bearer/example_test.go [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+
+	// init bearerToken, sdkClient, signer, other steps ...
+
+	var headParams client.PrmObjectHead
+	headParams.WithBearerToken(bearerToken)
+	// ...
+
+	response, err := sdkClient.ObjectHead(context.Background(), cid.ID{}, oid.ID{}, signer, headParams)
+
+	_ = response
```
[the remark this answers is not in the document]
Comment (smallhive) [reply bearer/example_test.go:60]: Correct, linter doesn't make a difference between examples and real code in this question

File: checksum/example_test.go [hunk 25]
```diff
@@ -13,7 +13,10 @@ func ExampleCalculate() {
 	payload := []byte{0, 1, 2, 3, 4, 5, 6}
 	var cs Checksum
 
+	// sha256Sum contains SHA256 hash of the payload
```
[the remark this answers is not in the document]
Comment (smallhive) [reply checksum/example_test.go]: Updated

File: checksum/example_test.go [hunk 26]
```diff
@@ -13,7 +13,10 @@ func ExampleCalculate() {
...
 	Calculate(&cs, SHA256, payload)
+
+	// tzSum contains TZ hash of the payload
```
[the remark this answers is not in the document]
Comment (smallhive) [reply checksum/example_test.go]: Updated

File: container/example_test.go [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+	// optional
+	cnr.SetName("awesome container name")
+	cnr.SetCreationTime(time.Now())
+	// ...
+
+	// placement policy and replicas definition is required
+	var pp netmap.PlacementPolicy
+	pp.SetContainerBackupFactor(1)
+
+	var rd netmap.ReplicaDescriptor
```
[the remark this answers is not in the document]
Comment (smallhive) [reply container/example_test.go]: Updated

File: container/example_test.go [hunk 28]
```diff
@@ -0,0 +1,64 @@
...
+	rd.SetNumberOfObjects(1)
+	pp.AddReplicas(rd)
+
+	cnr.SetPlacementPolicy(pp)
+}
+
+// Instances can be also used to process NeoFS API V2 protocol messages with [https://github.com/nspcc-dev/neofs-api] package.
+// On the client side.
+func ExampleContainer_WriteToV2() {
```
[the remark this answers is not in the document]
Comment (smallhive) [reply container/example_test.go]: Merged examples

**Review by roman-khimov — APPROVED** [review]

**Review by carpawell — APPROVED** [review]

**Pull request merged**